### DISCUSSION TITLE
Hugo live reload issue with Helix backup (.bck) files

### DISCUSSION THREAD

**Discussion by faresbakhit** [helix-editor/helix#11715] [General] [answered]
## Suggestion

Can an option be added to disable backup files and modify the files directly when saving changes with the `write:*` family of commands?

## Current Behavior

The current behavior can be tested by invoking `inotifywait -m .` from `inotify-tools`, opening a file, modifying the file, and saving the change with `:w`.

Output from `inotifywait` on my machine using `helix 24.7 (079f5442)`:

```
./ OPEN example.txt
./ ACCESS example.txt
./ CLOSE_NOWRITE,CLOSE example.txt
./ MOVED_FROM example.txt
./ MOVED_TO example.txtU77clA.bck
./ CREATE example.txt
./ OPEN example.txt
./ MODIFY example.txt
./ CLOSE_WRITE,CLOSE example.txt
./ ATTRIB example.txt
./ DELETE example.txtU77clA.bck
```

tl;dr: Helix reads `example.txt` contents into memory, and upon saving it moves (not copy) `example.txt` to `example.txtXXXXXX.bck`, creates `example.txt`, saves the new changes to `example.txt`, and removes `example.txtXXXXXX.bck`.

I'm not sure why Helix need to do this and the code doesn't help me either:

https://github.com/helix-editor/helix/blob/84fbadbdde255d87e6e04141ccd2d75f4e86edf2/helix-view/src/document.rs#L940-L970

## Use Case

Many tools that monitor a set of files/directories for changes and does *something* when files are created/changed/removed and don't produce the intended behavior since they get multiple events (`CREATE -> CREATE -> REMOVE`) instead of one `MODIFY` event.

Such tools include `latexmk`, `hugo`, and other static site generators or 'live' coding environments (although, I only use `latexmk` and `hugo`)

This is output from `hugo server --disableFastRender`:

```
Template added /partials/head.html5iipqu.bck
Template added /partials/head.html
```

And as a result it doesn't update the generated files properly. With the new option the output should be:

```
Template changed /partials/head.html
```

**Comment by pascalkuthe** [comment]
It's inktify 101 that you need to properly track files and can't just assume that files will be modified woth a single read. Editors like vim do that for example. It's the only way to avoid corrupting files if the editor crashes.

If these tools don't properly account for these schemes then their file watcher implementation is broken

**Comment by David-Else** [comment]
@pascalkuthe I have experienced the problem with Hugo and Helix myself, and confirmed it does not happen when using Vim. Do you know why this could be? Here is the issue in Hugo, @bep is the Hugo maintainer: https://github.com/gohugoio/hugo/issues/12677#issuecomment-2250443899

**Reply by pascalkuthe** [reply]
they hardcoded vim file names instead of implement a robust/general solution

**Comment by David-Else** [comment, marked as answer]
I added a fix to Hugo: https://github.com/gohugoio/hugo/pull/12856

EDIT: It was merged, should be fixed, please report if there are any more problems!

**Comment by faresbakhit** [comment]
@David-Else Thank you! I'll try compiling Hugo later and test it, but the patch should suffice.

As the bug was on Hugo's end, as @pascalkuthe said, I'll rename the discussion to reflect that so anyone with the same issue can just search for Hugo and upgrade to latest version with the patch.

**Comment by the-dipsy** [comment]
Unfortunately the safe writing didn't just trigger problems with the old version of Hugo but in all kinds of places. The [Parcel](https://github.com/parcel-bundler/parcel) bundler`s development server also breaks with the latest helix release. Changes made to files that are mounted into podman/docker containers will not update inside the container when the original file is replaced instead of overwritten as well.

I propose to introduce a `write-backup` option similar to (neo)vim's [`writebackup` option](https://neovim.io/doc/user/options.html#'writebackup') as I have [attempted in this branch](https://github.com/helix-editor/helix/compare/master...the-dipsy:helix:write-backup-option). If there are no objections, I'd like to create a PR.

**Reply by David-Else** [reply]
Good news, it has been done, it just needs to be merged: https://github.com/helix-editor/helix/pull/11374
